src/circuit_optimizer/CircuitOptimizer.cpp

```diff
@@ -1625,6 +1625,241 @@ void CircuitOptimizer::collectBlocks(QuantumComputation& qc,
   removeIdentities(qc);
 }
 
+/**
+ * @brief Block of Clifford operations
+ * @details This structure is used to collect Clifford operations that can be
+ * grouped together. It maintains the qubits that are part of the block
+ * and the qubits that are blocked by non-Clifford operations.
+ */
+struct CliffordBlock {
+  std::unordered_set<Qubit> blockQubits;
+  std::unordered_set<Qubit> blocked;
+  std::unique_ptr<CompoundOperation> operations;
+  std::unique_ptr<Operation>* position = nullptr;
+  std::size_t logicalStep = 0;
+
+  [[nodiscard]] bool empty() const noexcept {
+    return !operations || operations->empty();
+  }
+
+  /**
+   * @brief Check if this block is fully disabled by non-Clifford operations
+   *
+   */
+  [[nodiscard]] bool fullyDisabled() const noexcept {
+    return std::ranges::all_of(
+        blockQubits, [this](const Qubit q) { return blocked.contains(q); });
+  }
+
+  /**
+   * @brief Check if adding qubits used by gate would exceed maxBlockSize
+   * @param used Qubits used by the gate
+   * @param maxBlockSize Maximum allowed block size
+   */
+  [[nodiscard]] bool
+  checkMaxBlockSize(const std::set<Qubit>& used,
```

Comment:
The name of this function could be improved to reflect what it actually means to return true or false.